File: reservation/src/reservation/api/reservations/dto/reservation.py
from __future__ import annotations

from datetime import date
from uuid import UUID

from reservation.api.common import CamelModel
from reservation.domain import BookCondition, Reservation, ReservationStatus


class ReservationResponse(CamelModel):
    reservation_uid: UUID
    book_uid: UUID
    library_uid: UUID
    status: ReservationStatus
    start_date: date
    till_date: date
    condition_at_rent: BookCondition

    @classmethod
    def of(cls, reservation: Reservation) -> ReservationResponse:
        return cls.model_validate(reservation, from_attributes=True)

File: reservation/src/reservations/domain/reservation.ts
import { z } from 'zod';

export const RESERVATION_STATUSES = ['RENTED', 'RETURNED', 'EXPIRED'] as const;

export type ReservationStatus = (typeof RESERVATION_STATUSES)[number];

export interface Reservation {
  readonly reservationUid: string;
  readonly username: string;
  readonly bookUid: string;
  readonly libraryUid: string;
  readonly status: ReservationStatus;
  readonly startDate: Date;
  readonly tillDate: Date;
}

export const reservationSchema: z.ZodType<Reservation> = z.object({
  reservationUid: z.uuid(),
  username: z.string().trim().min(1).max(80),
  bookUid: z.uuid(),
  libraryUid: z.uuid(),
  status: z.enum(RESERVATION_STATUSES),
  startDate: z.coerce.date(),
  tillDate: z.coerce.date(),
});

export function validateReservation(value: unknown): Reservation {
  return reservationSchema.parse(value);
}
